sympy/combinatorics/homomorphisms.py

```diff
@@ -199,29 +244,42 @@ def homomorphism(domain, codomain, gens, images=[]):
     gens.extend([g for g in generators if g not in gens])
     images = dict(zip(gens,images))
 
-    if not _check_homomorphism(domain, images, codomain.identity):
+    if check and not _check_homomorphism(domain, codomain, images):
         raise ValueError("The given images do not define a homomorphism")
     return GroupHomomorphism(domain, codomain, images)
 
-def _check_homomorphism(domain, images, identity):
+def _check_homomorphism(domain, codomain, images):
     rels = domain.relators
+    identity = codomain.identity
     def _image(r):
         if r.is_identity:
             return identity
         else:
             w = identity
             r_arr = r.array_form
             i = 0
+            j = 0
             while i < len(r):
```

Comment:
I think a comment should be added to explain why both forms of `r` are needed. (Maybe the  definition  of array form should be changed to consist of pairs `(gen, exp)`.)